charmcraft/application/main.py

```diff
@@ -0,0 +1,105 @@
+# Copyright 2023 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For further info, check https://github.com/canonical/charmcraft
+"""New entrypoint for charmcraft."""
+from __future__ import annotations
+
+import signal
+import sys
+
+import craft_cli
+from craft_application import Application, AppMetadata, ServiceFactory
+from craft_cli import emit
+
+from charmcraft import errors, models
+from charmcraft.main import GENERAL_SUMMARY
+from charmcraft.main import main as old_main
+
+APP_METADATA = AppMetadata(
+    name="charmcraft",
+    summary=GENERAL_SUMMARY,
+    ProjectClass=models.charmcraft.Project,
+)
+
+
+class Charmcraft(Application):
+    """Charmcraft application definition."""
+
+    @property
+    def command_groups(self) -> list[craft_cli.CommandGroup]:
+        """Excluding lifecycle commands for right now."""
+        return self._command_groups
+
+    def _configure_services(self) -> None:
+        super()._configure_services()
+        self.services.set_kwargs(
+            "package",
+            work_dir=self._work_dir,
+            prime_dir=self.services.lifecycle.prime_dir,
+        )
+
+    def _get_dispatcher(self) -> craft_cli.Dispatcher:
+        """Configure charmcraft, including a fallback to the classic entrypoint.
+
+        Side-effect: This method may exit the process.
+        Raises: ClassicFallback() to fall back to the classic interface.
```

Comment:
Is this style on purpose?  I expected `:raises ClassicFallback: <explanation>`.

Reply:
Nope! Not intentional. Good catch!